wolfcrypt/src/port/psa/psa_engine.c

```diff
@@ -0,0 +1,704 @@
+/* psa.c
+ *
+ * Copyright (C) 2006-2021 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+#ifdef HAVE_CONFIG_H
+    #include <config.h>
+#endif
+
+#include <wolfssl/wolfcrypt/settings.h>
+
+#if defined(WOLFSSL_HAVE_PSA_ENGINE)
+
+#include <wolfssl/wolfcrypt/port/psa/crypto.h>
+
+#if defined(HAVE_ECC)
+#include <wolfssl/wolfcrypt/ecc.h>
+#endif
+
+#if !defined(NO_SHA256)
+#include <wolfssl/wolfcrypt/sha256.h>
+#endif
+
+#if !defined(NO_AES)
+#include <wolfssl/wolfcrypt/aes.h>
+#endif
+
+#include <wolfssl/wolfcrypt/random.h>
+
+static psa_aead_operation_t zero_aead_operation;
+static psa_cipher_operation_t zero_cipher_operation;
+static psa_hash_operation_t zero_hash_operation;
+static psa_key_derivation_operation_t zero_key_derivation_operation;
+static psa_mac_operation_t zero_mac_operation;
+static psa_key_attributes_t zero_key_attribute;
+
+static int psa_initialized;
+static psa_key_id_t psa_incremental_id;
+
+/* max number of keys that can be used */
+#define PSA_KEY_SLOTS 6
+
+/** struct wc_psa_key
+ * @attr: attributes of a key
+ * @key: wolfssl native key used
+ */
+struct wc_psa_key {
+    psa_key_attributes_t attr;
+    void *key;
+};
+
+static wolfSSL_Mutex psa_key_slots_lock;
+static struct wc_psa_key wc_psa_key_slots[PSA_KEY_SLOTS];
+
+static struct wc_psa_key *psa_find_key(psa_key_id_t id)
+{
+    struct wc_psa_key *k;
+    int i;
+
+    if (id == PSA_KEY_ID_NULL)
+        return NULL;
+
+    if (id < PSA_KEY_ID_USER_MIN || id > PSA_KEY_ID_USER_MAX)
+        return NULL;
+
+    for (i = 0; i < PSA_KEY_SLOTS; ++i) {
+        k = &wc_psa_key_slots[i];
+        if (k->attr.id == id)
+            return k;
+    }
+
+    return NULL;
+}
+
+static psa_key_id_t psa_get_new_id(void)
+{
+    int err;
+
+    err = wc_LockMutex(&psa_key_slots_lock);
+    if (err != 0)
+        return PSA_KEY_ID_NULL;
+    psa_incremental_id++;
+    err = wc_UnLockMutex(&psa_key_slots_lock);
+
+    if (psa_incremental_id > PSA_KEY_ID_USER_MAX)
```

Comment:
I  think it's better to move this check in the inside the lock protected area, so we can avoid to increase `psa_incremental_id` above `PSA_KEY_ID_USER_MAX`